File: go/protocols/mpp/intents/session.go
package intents

import (
	"encoding/json"
	"fmt"
	"strconv"

	solana "github.com/gagliardetto/solana-go"

	"github.com/solana-foundation/pay-kit/go/protocols/mpp/program"
)

// DefaultSessionExpiresAt is the default session voucher/directive expiry:
// 2100-01-01T00:00:00Z. It stays below JavaScript's max safe integer so JSON
// intermediaries do not round it before the credential is decoded. Mirrors
// DEFAULT_SESSION_EXPIRES_AT in the Rust spine.
const DefaultSessionExpiresAt int64 = 4_102_444_800

// SessionMode is the on-chain funding mechanism for a session.
type SessionMode string

const (
	// SessionModePush is a payment channel backed by an on-chain escrow
	// deposit (client-funded).
	SessionModePush SessionMode = "push"
	// SessionModePull is an operator-assisted session whose voucher authority
	// is declared separately via SessionPullVoucherStrategy.
	SessionModePull SessionMode = "pull"
)

// SessionPullVoucherStrategy is the voucher authority used when pull mode is
// advertised.
type SessionPullVoucherStrategy string

const (
	// PullStrategyClientVoucher means the client signs cumulative vouchers.
	PullStrategyClientVoucher SessionPullVoucherStrategy = "clientVoucher"
	// PullStrategyOperatedVoucher means the operator signs vouchers.
	PullStrategyOperatedVoucher SessionPullVoucherStrategy = "operatedVoucher"
)

// SessionSplit is a payment split committed at channel open; distributed to a
// specific recipient when the channel closes.
type SessionSplit struct {
	Recipient string `json:"recipient"`
	Bps       uint16 `json:"bps"`
}

// SessionRequest is the session intent request embedded in a 402 challenge.
// Field casing and omit rules mirror SessionRequest in the Rust spine.
type SessionRequest struct {
	Cap                 string                      `json:"cap"`
	Currency            string                      `json:"currency"`
	Decimals            *uint8                      `json:"decimals,omitempty"`
	Network             string                      `json:"network,omitempty"`
	Operator            string                      `json:"operator"`
	Recipient           string                      `json:"recipient"`
	Splits              []SessionSplit              `json:"splits,omitempty"`
	ProgramID           string                      `json:"programId,omitempty"`
	Description         string                      `json:"description,omitempty"`
	ExternalID          string                      `json:"externalId,omitempty"`
	MinVoucherDelta     string                      `json:"minVoucherDelta,omitempty"`
	Modes               []SessionMode               `json:"modes,omitempty"`
	PullVoucherStrategy *SessionPullVoucherStrategy `json:"pullVoucherStrategy,omitempty"`
	RecentBlockhash     string                      `json:"recentBlockhash,omitempty"`
}

// OpenPayload is the payload for the open action. Shape varies by mode; use
// the constructors to build it and Mode to distinguish variants on the server.
type OpenPayload struct {
	Mode SessionMode `json:"mode"`

	// Push mode.
	ChannelID   string  `json:"channelId,omitempty"`
	Deposit     string  `json:"deposit,omitempty"`
	Payer       string  `json:"payer,omitempty"`
	Payee       string  `json:"payee,omitempty"`
	Mint        string  `json:"mint,omitempty"`
	Salt        *uint64 `json:"-"`
	GracePeriod *uint32 `json:"gracePeriod,omitempty"`
	Transaction string  `json:"transaction,omitempty"`

	// Pull mode.
	TokenAccount        string `json:"tokenAccount,omitempty"`
	ApprovedAmount      string `json:"approvedAmount,omitempty"`
	Owner               string `json:"owner,omitempty"`
	InitMultiDelegateTx string `json:"initMultiDelegateTx,omitempty"`
	UpdateDelegationTx  string `json:"updateDelegationTx,omitempty"`

	// Shared.
	AuthorizedSigner string `json:"authorizedSigner"`
	Signature        string `json:"signature"`
}

// openPayloadWire is the JSON projection of OpenPayload with salt serialized
// as a decimal string (number-tolerant on read). JSON numbers > 2^53 are
// unsafe in JS intermediaries, so salt is always emitted as a string.
type openPayloadWire struct {
	Mode                SessionMode      `json:"mode"`
	ChannelID           string           `json:"channelId,omitempty"`
	Deposit             string           `json:"deposit,omitempty"`
	Payer               string           `json:"payer,omitempty"`
	Payee               string           `json:"payee,omitempty"`
	Mint                string           `json:"mint,omitempty"`
	Salt                *string          `json:"salt,omitempty"`
	GracePeriod         *uint32          `json:"gracePeriod,omitempty"`
	Transaction         string           `json:"transaction,omitempty"`
	TokenAccount        string           `json:"tokenAccount,omitempty"`
	ApprovedAmount      string           `json:"approvedAmount,omitempty"`
	Owner               string           `json:"owner,omitempty"`
	InitMultiDelegateTx string           `json:"initMultiDelegateTx,omitempty"`
	UpdateDelegationTx  string           `json:"updateDelegationTx,omitempty"`
	AuthorizedSigner    string           `json:"authorizedSigner"`
	Signature           string           `json:"signature"`
	SaltRaw             *json.RawMessage `json:"-"`
}

// MarshalJSON emits salt as a decimal string and omits empty/None fields.
func (p OpenPayload) MarshalJSON() ([]byte, error) {
	w := openPayloadWire{
		Mode:                p.Mode,
		ChannelID:           p.ChannelID,
		Deposit:             p.Deposit,
		Payer:               p.Payer,
		Payee:               p.Payee,
		Mint:                p.Mint,
		GracePeriod:         p.GracePeriod,
		Transaction:         p.Transaction,
		TokenAccount:        p.TokenAccount,
		ApprovedAmount:      p.ApprovedAmount,
		Owner:               p.Owner,
		InitMultiDelegateTx: p.InitMultiDelegateTx,
		UpdateDelegationTx:  p.UpdateDelegationTx,
		AuthorizedSigner:    p.AuthorizedSigner,
		Signature:           p.Signature,
	}
	if p.Salt != nil {
		s := strconv.FormatUint(*p.Salt, 10)
		w.Salt = &s
	}
	return json.Marshal(w)
}

// UnmarshalJSON reads salt from either a decimal string or a JSON number.
func (p *OpenPayload) UnmarshalJSON(data []byte) error {
	var w struct {
		openPayloadWire
		Salt json.RawMessage `json:"salt"`
	}
	if err := json.Unmarshal(data, &w); err != nil {
		return err
	}
	p.Mode = w.Mode
	p.ChannelID = w.ChannelID
	p.Deposit = w.Deposit
	p.Payer = w.Payer
	p.Payee = w.Payee
	p.Mint = w.Mint
	p.GracePeriod = w.GracePeriod
	p.Transaction = w.Transaction
	p.TokenAccount = w.TokenAccount
	p.ApprovedAmount = w.ApprovedAmount
	p.Owner = w.Owner
	p.InitMultiDelegateTx = w.InitMultiDelegateTx
	p.UpdateDelegationTx = w.UpdateDelegationTx
	p.AuthorizedSigner = w.AuthorizedSigner
	p.Signature = w.Signature
	salt, err := parseOptionalU64(w.Salt)
	if err != nil {
		return err
	}
	p.Salt = salt
	return nil
}

// parseOptionalU64 accepts a decimal string, a JSON number, or null.
func parseOptionalU64(raw json.RawMessage) (*uint64, error) {
	if len(raw) == 0 || string(raw) == "null" {
		return nil, nil
	}
	if raw[0] == '"' {
		var s string
		if err := json.Unmarshal(raw, &s); err != nil {
			return nil, err
		}
		value, err := strconv.ParseUint(s, 10, 64)
		if err != nil {
			return nil, fmt.Errorf("salt must be a decimal string: %w", err)
		}
		return &value, nil
	}
	value, err := strconv.ParseUint(string(raw), 10, 64)
	if err != nil {
		return nil, fmt.Errorf("salt must be an unsigned 64-bit integer: %w", err)
	}
	return &value, nil
}

// NewOpenPush builds a push payment-channel open payload.
func NewOpenPush(channelID, deposit, authorizedSigner, signature string) OpenPayload {
	return OpenPayload{
		Mode:             SessionModePush,
		ChannelID:        channelID,
		Deposit:          deposit,
		AuthorizedSigner: authorizedSigner,
		Signature:        signature,
	}
}

// NewOpenPaymentChannel builds a push payment-channel open payload carrying the
// full channel parameters.
func NewOpenPaymentChannel(
	channelID, deposit, payer, payee, mint string,
	salt uint64, gracePeriod uint32,
	authorizedSigner, signature string,
) OpenPayload {
	return NewOpenPaymentChannelWithMode(SessionModePush, channelID, deposit, payer, payee, mint, salt, gracePeriod, authorizedSigner, signature)
}

// NewOpenPaymentChannelWithMode builds a payment-channel open payload with an
// explicit submission mode (push: client broadcasts; pull: operator broadcasts).
func NewOpenPaymentChannelWithMode(
	mode SessionMode,
	channelID, deposit, payer, payee, mint string,
	salt uint64, gracePeriod uint32,
	authorizedSigner, signature string,
) OpenPayload {
	saltCopy := salt
	graceCopy := gracePeriod
	return OpenPayload{
		Mode:             mode,
		ChannelID:        channelID,
		Deposit:          deposit,
		Payer:            payer,
		Payee:            payee,
		Mint:             mint,
		Salt:             &saltCopy,
		GracePeriod:      &graceCopy,
		AuthorizedSigner: authorizedSigner,
		Signature:        signature,
	}
}

// NewOpenPull builds a pull (SPL delegation) open payload.
func NewOpenPull(tokenAccount, approvedAmount, owner, authorizedSigner, signature string) OpenPayload {
	return OpenPayload{
		Mode:             SessionModePull,
		TokenAccount:     tokenAccount,
		ApprovedAmount:   approvedAmount,
		Owner:            owner,
		AuthorizedSigner: authorizedSigner,
		Signature:        signature,
	}
}

// SessionID returns the session store key: channelId for push, tokenAccount
// for pull-without-channel.
func (p OpenPayload) SessionID() (string, error) {
	if p.ChannelID != "" {
		return p.ChannelID, nil
	}
	switch p.Mode {
	case SessionModePush:
		return "", fmt.Errorf("push open missing channelId")
	case SessionModePull:
		if p.TokenAccount == "" {
			return "", fmt.Errorf("pull open missing channelId or tokenAccount")
		}
		return p.TokenAccount, nil
	default:
		return "", fmt.Errorf("unknown session mode %q", p.Mode)
	}
}

// DepositAmount returns the deposit (push) or approved amount (pull) in base units.
func (p OpenPayload) DepositAmount() (uint64, error) {
	raw := p.Deposit
	if raw == "" {
		switch p.Mode {
		case SessionModePush:
			return 0, fmt.Errorf("push open missing deposit")
		case SessionModePull:
			if p.ApprovedAmount == "" {
				return 0, fmt.Errorf("pull open missing deposit or approvedAmount")
			}
			raw = p.ApprovedAmount
		default:
			return 0, fmt.Errorf("unknown session mode %q", p.Mode)
		}
	}
	value, err := strconv.ParseUint(raw, 10, 64)
	if err != nil {
		return 0, fmt.Errorf("invalid deposit amount: %s", raw)
	}
	return value, nil
}

// VoucherData is the canonical content of a voucher signed by the client's
// session key. The wire JSON carries channelId (base58), cumulativeAmount
// (string; also accepts the legacy "cumulative" alias on read), and expiresAt
// (i64). The signed bytes are the program Borsh VoucherArgs layout.
type VoucherData struct {
	ChannelID  string
	Cumulative string
	ExpiresAt  int64
	Nonce      *uint64
}

type voucherDataWire struct {
	ChannelID        string  `json:"channelId"`
	CumulativeAmount string  `json:"cumulativeAmount"`
	Cumulative       string  `json:"cumulative,omitempty"`
	ExpiresAt        int64   `json:"expiresAt"`
	Nonce            *uint64 `json:"nonce,omitempty"`
}

// MarshalJSON serializes cumulative only as cumulativeAmount.
func (v VoucherData) MarshalJSON() ([]byte, error) {
	return json.Marshal(voucherDataWire{
		ChannelID:        v.ChannelID,
		CumulativeAmount: v.Cumulative,
		ExpiresAt:        v.ExpiresAt,
		Nonce:            v.Nonce,
	})
}

// UnmarshalJSON accepts both cumulativeAmount and the legacy cumulative alias.
func (v *VoucherData) UnmarshalJSON(data []byte) error {
	var w voucherDataWire
	if err := json.Unmarshal(data, &w); err != nil {
		return err
	}
	v.ChannelID = w.ChannelID
	v.Cumulative = w.CumulativeAmount
	if v.Cumulative == "" {
		v.Cumulative = w.Cumulative
	}
	v.ExpiresAt = w.ExpiresAt
	v.Nonce = w.Nonce
	return nil
}

// MessageBytes serializes the voucher to the payment-channels VoucherArgs bytes
// signed by Ed25519: channelId(32) || cumulative(u64 LE) || expiresAt(i64 LE).
func (v VoucherData) MessageBytes() ([]byte, error) {
	channelID, err := solana.PublicKeyFromBase58(v.ChannelID)
	if err != nil {
		return nil, fmt.Errorf("invalid voucher channelId: %w", err)
	}
	cumulative, err := strconv.ParseUint(v.Cumulative, 10, 64)
	if err != nil {
		return nil, fmt.Errorf("invalid voucher cumulative: %s", v.Cumulative)
	}
	return program.VoucherMessageBytes(channelID, cumulative, v.ExpiresAt), nil
}

// SignedVoucher is a voucher signed by the client's session key. Vouchers are
// cumulative: the server always uses the latest valid voucher it received.
type SignedVoucher struct {
	Data      VoucherData `json:"data"`
	Signature string      `json:"signature"`
}

// VoucherPayload is the payload for the voucher action.
type VoucherPayload struct {
	Voucher SignedVoucher `json:"voucher"`
}

// CommitPayload is the payload for the commit action.
type CommitPayload struct {
	DeliveryID string        `json:"deliveryId"`
	Voucher    SignedVoucher `json:"voucher"`
}

// TopUpPayload is the payload for the topup action.
type TopUpPayload struct {
	ChannelID  string `json:"channelId"`
	NewDeposit string `json:"newDeposit"`
	Signature  string `json:"signature"`
}

// ClosePayload is the payload for the close action.
type ClosePayload struct {
	ChannelID string         `json:"channelId"`
	Voucher   *SignedVoucher `json:"voucher,omitempty"`
}

// SessionActionTag identifies the action variant.
type SessionActionTag string

const (
	// ActionOpen opens a new channel/delegation and starts the session.
	ActionOpen SessionActionTag = "open"
	// ActionVoucher submits a signed voucher authorizing payment.
	ActionVoucher SessionActionTag = "voucher"
	// ActionCommit commits a metered delivery by attaching a signed voucher.
	ActionCommit SessionActionTag = "commit"
	// ActionTopUp tops up an existing channel's deposit. The wire tag uses a
	// capital U ("topUp") to match the Rust serde camelCase rename.
	ActionTopUp SessionActionTag = "topUp"
	// ActionClose requests cooperative close of the channel.
	ActionClose SessionActionTag = "close"
)

// SessionAction is the tagged action submitted by the client in an
// Authorization header, discriminated by the "action" field. Exactly one of
// the payload pointers is set, matching the active tag.
type SessionAction struct {
	Action  SessionActionTag
	Open    *OpenPayload
	Voucher *VoucherPayload
	Commit  *CommitPayload
	TopUp   *TopUpPayload
	Close   *ClosePayload
}

// MarshalJSON flattens the active payload alongside the action tag.
func (a SessionAction) MarshalJSON() ([]byte, error) {
	var payload any
	switch a.Action {
	case ActionOpen:
		payload = a.Open
	case ActionVoucher:
		payload = a.Voucher
	case ActionCommit:
		payload = a.Commit
	case ActionTopUp:
		payload = a.TopUp
	case ActionClose:
		payload = a.Close
	default:
		return nil, fmt.Errorf("unknown session action %q", a.Action)
	}
	if payload == nil {
		return nil, fmt.Errorf("session action %q has no payload", a.Action)
	}
	raw, err := json.Marshal(payload)
	if err != nil {
		return nil, err
	}
	var fields map[string]json.RawMessage
	if err := json.Unmarshal(raw, &fields); err != nil {
		return nil, err
	}
	tag, err := json.Marshal(string(a.Action))
	if err != nil {
		return nil, err
	}
	fields["action"] = tag
	return json.Marshal(fields)
}

// UnmarshalJSON reads the action tag then decodes the matching payload.
func (a *SessionAction) UnmarshalJSON(data []byte) error {
	var tag struct {
		Action SessionActionTag `json:"action"`
	}
	if err := json.Unmarshal(data, &tag); err != nil {
		return err
	}
	a.Action = tag.Action
	switch tag.Action {
	case ActionOpen:
		var p OpenPayload
		if err := json.Unmarshal(data, &p); err != nil {
			return err
		}
		a.Open = &p
	case ActionVoucher:
		var p VoucherPayload
		if err := json.Unmarshal(data, &p); err != nil {
			return err
		}
		a.Voucher = &p
	case ActionCommit:
		var p CommitPayload
		if err := json.Unmarshal(data, &p); err != nil {
			return err
		}
		a.Commit = &p
	case ActionTopUp:
		var p TopUpPayload
		if err := json.Unmarshal(data, &p); err != nil {
			return err
		}
		a.TopUp = &p
	case ActionClose:
		var p ClosePayload
		if err := json.Unmarshal(data, &p); err != nil {
			return err
		}
		a.Close = &p
	default:
		return fmt.Errorf("unknown session action %q", tag.Action)
	}
	return nil
}

// MeteringDirective is the server-issued directive attached to a delivered
// message/response. After processing, the client signs a voucher covering
// amount and commits it referencing deliveryId.
type MeteringDirective struct {
	DeliveryID string `json:"deliveryId"`
	SessionID  string `json:"sessionId"`
	Amount     string `json:"amount"`
	Currency   string `json:"currency"`
	Sequence   uint64 `json:"sequence"`
	ExpiresAt  int64  `json:"expiresAt"`
	CommitURL  string `json:"commitUrl,omitempty"`
	Proof      string `json:"proof,omitempty"`
}

// AmountBaseUnits parses the directive amount as base units.
func (d MeteringDirective) AmountBaseUnits() (uint64, error) {
	value, err := strconv.ParseUint(d.Amount, 10, 64)
	if err != nil {
		return 0, fmt.Errorf("invalid metering amount: %s", d.Amount)
	}
	return value, nil
}

// MeteringUsage is the final usage reported by a streaming response. The amount
// must be less than or equal to the amount reserved by the directive.
type MeteringUsage struct {
	DeliveryID string `json:"deliveryId"`
	Amount     string `json:"amount"`
}

// AmountBaseUnits parses the usage amount as base units.
func (u MeteringUsage) AmountBaseUnits() (uint64, error) {
	value, err := strconv.ParseUint(u.Amount, 10, 64)
	if err != nil {
		return 0, fmt.Errorf("invalid metering usage amount: %s", u.Amount)
	}
	return value, nil
}

// MeteredEnvelope pairs a payload with the metering directive needed to
// acknowledge it.
type MeteredEnvelope[T any] struct {
	Payload  T                 `json:"payload"`
	Metering MeteringDirective `json:"metering"`
}

// CommitStatus is the commit receipt status.
type CommitStatus string

const (
	// CommitStatusCommitted is the first successful commit for a delivery.
	CommitStatusCommitted CommitStatus = "committed"
	// CommitStatusReplayed is an idempotent replay of a prior accepted commit.
	CommitStatusReplayed CommitStatus = "replayed"
)

// CommitReceipt is returned after a delivery commit is accepted.
type CommitReceipt struct {
	DeliveryID string       `json:"deliveryId"`
	SessionID  string       `json:"sessionId"`
	Amount     string       `json:"amount"`
	Cumulative string       `json:"cumulative"`
	Status     CommitStatus `json:"status"`
}

File: lua/pay_kit/protocols/mpp/session.lua
--[[
MPP session intent wire types.

Mirrors the Rust spine at `rust/crates/mpp/src/protocol/intents/session.rs`.
The session intent opens a payment channel so a client can pay incrementally
with off-chain signed cumulative vouchers, settled on-chain only at
open / top-up / close.

LOAD-BEARING WIRE PARITY (must match Rust + the TypeScript package):

  - `cumulativeAmount` is the wire field name; `cumulative` is accepted as a
    read alias on decode. Serialization always emits `cumulativeAmount`.
  - `salt` serializes as a decimal STRING (JSON numbers > 2^53 are unsafe in
    JS intermediaries) but decode tolerates both string and number.
  - The `topup` action tag is `topUp` (capital U) on the wire.
  - `DEFAULT_SESSION_EXPIRES_AT == 4102444800` (2100-01-01 UTC), chosen to
    stay below Number.MAX_SAFE_INTEGER.
  - Voucher SIGNING bytes are Borsh, not JSON: the JSON carries channelId
    (base58), cumulativeAmount (string), expiresAt (i64); the signed bytes
    are channelId(32) || cumulative(u64 LE) || expiresAt(i64 LE). See
    `pay_kit.solana.payment_channels.voucher_message_bytes`.

This module is pure data shaping (encode/decode helpers + constructors). The
server lifecycle lives in `pay_kit.protocols.mpp.server.session_handler`.
]]

local payment_channels = require('pay_kit.solana.payment_channels')

local M = {}

-- 2100-01-01T00:00:00Z. Below Number.MAX_SAFE_INTEGER on purpose.
M.DEFAULT_SESSION_EXPIRES_AT = 4102444800

-- ── Session modes ──

M.MODE_PUSH = 'push'
M.MODE_PULL = 'pull'

M.PULL_STRATEGY_CLIENT_VOUCHER = 'clientVoucher'
M.PULL_STRATEGY_OPERATED_VOUCHER = 'operatedVoucher'

-- ── Commit status ──

M.COMMIT_STATUS_COMMITTED = 'committed'
M.COMMIT_STATUS_REPLAYED = 'replayed'

-- Strip nil-valued fields so encoders that walk pairs() do not emit them.
-- The Rust spine uses `skip_serializing_if = Option::is_none`; this is the
-- Lua equivalent. (We build tables with only present keys, so this is mostly
-- a guard for callers that pass explicit nils.)
local function prune(tbl)
  for k, v in pairs(tbl) do
    if v == nil then
      tbl[k] = nil
    end
  end
  return tbl
end

--- Build a SessionRequest table (the payload embedded in a 402 challenge).
-- Required: cap (string), currency, operator, recipient.
-- Optional: decimals, network, splits, program_id, description, external_id,
-- min_voucher_delta, modes, pull_voucher_strategy, recent_blockhash.
-- Empty splits/modes and nil optionals are omitted, matching the Rust
-- `skip_serializing_if` behavior.
function M.session_request(opts)
  opts = opts or {}
  local req = { cap = tostring(opts.cap), currency = opts.currency }
  req.operator = opts.operator
  req.recipient = opts.recipient
  if opts.decimals ~= nil then req.decimals = opts.decimals end
  if opts.network ~= nil then req.network = opts.network end
  if opts.splits and #opts.splits > 0 then req.splits = opts.splits end
  if opts.program_id ~= nil then req.programId = opts.program_id end
  if opts.description ~= nil then req.description = opts.description end
  if opts.external_id ~= nil then req.externalId = opts.external_id end
  if opts.min_voucher_delta ~= nil then req.minVoucherDelta = tostring(opts.min_voucher_delta) end
  if opts.modes and #opts.modes > 0 then req.modes = opts.modes end
  if opts.pull_voucher_strategy ~= nil then req.pullVoucherStrategy = opts.pull_voucher_strategy end
  if opts.recent_blockhash ~= nil then req.recentBlockhash = opts.recent_blockhash end
  return req
end

-- ── OpenPayload ──

--- Construct a push (payment-channel) open payload from a signed-tx signature.
function M.open_push(channel_id, deposit, authorized_signer, signature)
  return {
    action = 'open',
    mode = M.MODE_PUSH,
    channelId = channel_id,
    deposit = tostring(deposit),
    authorizedSigner = authorized_signer,
    signature = signature,
  }
end

--- Construct a full payment-channel push open payload (carries payer/payee/
--- mint/salt/gracePeriod). `mode` defaults to push.
function M.open_payment_channel(opts)
  local p = {
    action = 'open',
    mode = opts.mode or M.MODE_PUSH,
    channelId = opts.channel_id,
    deposit = tostring(opts.deposit),
    payer = opts.payer,
    payee = opts.payee,
    mint = opts.mint,
    salt = M.encode_salt(opts.salt),
    gracePeriod = opts.grace_period,
    authorizedSigner = opts.authorized_signer,
    signature = opts.signature,
  }
  if opts.transaction ~= nil then p.transaction = opts.transaction end
  return prune(p)
end

--- Construct a pull (operated-voucher SPL delegation) open payload.
function M.open_pull(token_account, approved_amount, owner, authorized_signer, signature)
  return {
    action = 'open',
    mode = M.MODE_PULL,
    tokenAccount = token_account,
    approvedAmount = tostring(approved_amount),
    owner = owner,
    authorizedSigner = authorized_signer,
    signature = signature,
  }
end

--- `salt` always serializes as a decimal string. Accepts number or string.
function M.encode_salt(salt)
  if salt == nil then return nil end
  return tostring(salt)
end

--- Tolerant salt decode: accepts string or number, returns a decimal string
--- (preserving u64 precision past 2^53 when the input is already a string).
function M.decode_salt(salt)
  if salt == nil then return nil end
  if type(salt) == 'number' then
    return string.format('%.0f', salt)
  end
  return tostring(salt)
end

--- The session id used as the channel-store key.
--- Push: channelId. Pull-no-channel: tokenAccount.
function M.session_id(payload)
  if payload.channelId and payload.channelId ~= '' then
    return payload.channelId
  end
  if payload.mode == M.MODE_PULL and payload.tokenAccount and payload.tokenAccount ~= '' then
    return payload.tokenAccount
  end
  error('open payload missing channelId or tokenAccount')
end

--- Deposit / approved amount for an open, as a decimal string.
function M.deposit_amount(payload)
  local raw = payload.deposit
  if raw == nil and payload.mode == M.MODE_PULL then
    raw = payload.approvedAmount
  end
  if raw == nil or tostring(raw) == '' then
    error('open payload missing deposit or approvedAmount')
  end
  raw = tostring(raw)
  if not raw:match('^%d+$') then
    error('invalid deposit amount: ' .. raw)
  end
  return raw
end

-- ── Vouchers ──

--- Read the cumulative amount from a voucher's data, accepting either the
--- `cumulativeAmount` wire name or the `cumulative` read alias.
function M.voucher_cumulative(data)
  local raw = data.cumulativeAmount
  if raw == nil then raw = data.cumulative end
  if raw == nil then
    error('voucher data missing cumulativeAmount')
  end
  return tostring(raw)
end

--- Serialize a VoucherData table to the canonical wire shape. Always emits
--- `cumulativeAmount` (never the `cumulative` alias).
function M.voucher_data(channel_id, cumulative_amount, expires_at, nonce)
  local data = {
    channelId = channel_id,
    cumulativeAmount = tostring(cumulative_amount),
    expiresAt = expires_at,
  }
  if nonce ~= nil then data.nonce = nonce end
  return data
end

--- The 48-byte Ed25519 signing message for a voucher data table.
function M.voucher_message_bytes(data)
  return payment_channels.voucher_message_bytes(
    data.channelId,
    M.voucher_cumulative(data),
    data.expiresAt
  )
end

-- ── Action payload constructors (client-side; useful for tests/harness) ──

function M.voucher_action(signed_voucher)
  return { action = 'voucher', voucher = signed_voucher }
end

function M.commit_action(delivery_id, signed_voucher)
  return { action = 'commit', deliveryId = delivery_id, voucher = signed_voucher }
end

function M.topup_action(channel_id, new_deposit, signature)
  return {
    action = 'topUp',
    channelId = channel_id,
    newDeposit = tostring(new_deposit),
    signature = signature,
  }
end

function M.close_action(channel_id, signed_voucher)
  local a = { action = 'close', channelId = channel_id }
  if signed_voucher ~= nil then a.voucher = signed_voucher end
  return a
end

return M
